github-connector/internal/registration/registerservice.go

```diff
@@ -1,1020 +1,123 @@
+func NewServiceRegister(deploymentEnvName string, b Builder, retryTime int, retries int) serviceRegister {
+
+	return serviceRegister{
+		envName:      deploymentEnvName,
+		builder:      b,
+		retryDelay:   retryTime * int(time.Second),
+		retriesCount: retries,
+	}
+}
 
 //RegisterService - register service in Kyma and get a response
-func RegisterService() (string, apperrors.AppError) {
+func (r serviceRegister) RegisterService() (string, apperrors.AppError) {
+
+	jsonBody, err := r.builder.BuildServiceDetails()
+	if err != nil {
+		return "", apperrors.Internal("While building service details json: %s", err)
+	}
+	id, err := jsonBody.requestWithRetries(r.envName, r.builder.GetApplicationRegistryURL(), r.retryDelay, r.retriesCount)
+	if err != nil {
+		return "", apperrors.Internal("While trying to register service: %s", err.Error())
+	}
+	return id, nil
+}
 
+func (jsonBody *ServiceDetails) requestWithRetries(appName string, url string, retryDelay int, retriesCount int) (string, error) {
 	var id string
 	var err error
-	for i := 0; i < 10; i++ {
-		id, err = SendRegisterRequest(jsonBody, url)
+
+	for i := 0; i < retriesCount; i++ {
+		time.Sleep(time.Duration(retryDelay))
+		id, err = sendRequest(*jsonBody, url)
 		if err == nil {
 			break
 		}
 		log.Warn(err.Error())
+	}
+	if err != nil {
+		return "", apperrors.UpstreamServerCallFailed("While sending service registration request: %s", err)
+	}
+	return id, nil
+}
 
-		time.Sleep(5 * time.Second)
+//RegisterResponse contain structure of response json
+type RegisterResponse struct {
+	ID string
+}
+
+//Do - create request and send it to kyma application registry
+func sendRequest(JSONBody ServiceDetails, url string) (string, error) {
+
+	// parse json to io.Reader
+	requestByte, err := json.Marshal(JSONBody)
+	if err != nil {
+		return "", apperrors.Internal("Failed to parse application registry request JSON body: %s", err.Error())
 	}
 
+	requestReader := bytes.NewReader(requestByte)
+
+	httpRequest, err := http.NewRequest(http.MethodPost, url, requestReader)
+
 	if err != nil {
-		return "", apperrors.Internal("While trying to register service: %s", err.Error())
+		return "", apperrors.Internal("Failed to create JSON request: %s", err.Error())
 	}
-	return id, nil
+
+	httpRequest.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	httpResponse, err := client.Do(httpRequest)
+
+	if err != nil {
+		return "", apperrors.UpstreamServerCallFailed("Failed to make request to '%s': %s", url, err.Error())
+	}
+
+	if httpResponse.StatusCode != http.StatusOK {
+		return "", apperrors.UpstreamServerCallFailed("Incorrect response code '%d' while sending JSON request from %s", httpResponse.StatusCode, url)
+	}
+
+	bodyBytes, err := ioutil.ReadAll(httpResponse.Body)
+
+	if err != nil {
+		return "", apperrors.UpstreamServerCallFailed("Failed to read service ID from application registry JSON response: %s", err)
```

Comment:
In this case the call succeeded but the response differs from the expected. It could take place if the response format has been changed in the Application Registry. This is pretty unlikely situation: Application Registry is required to be backward compatible. I think InternalError is more suitable here.

Reply:
Good point, thanks!